packages/runner-ct/src/iframe/iframes.tsx

```diff
@@ -22,134 +21,133 @@ export function getSpecUrl ({ namespace, spec }, prefix = '') {
-          className={
-            cs('size-container', {
-              [styles.noSpecAut]: !this.props.state.spec,
-            })
-          }
-          style={{
-            height: viewportHeight,
-            width: viewportWidth,
-            transform: `scale(${screenshotting ? 1 : scale})`,
-          }}
+export const Iframes: React.FC<IFramesProps> = namedObserver('Iframes', (props) => {
+  const containerRef = useRef<HTMLDivElement>(null)
+  const [autIframe] = useState(new AutIframe(props.config))
+  const [disposers, setDisposers] = useState<IReactionDisposer[]>([])
+
+  const [iframeModel] = useState(
+    new IframeModel({
+      state: props.state,
+      restoreDom: autIframe.restoreDom,
+      highlightEl: autIframe.highlightEl,
+      detachDom: autIframe.detachDom,
+      snapshotControls: (snapshotProps) => (
+        <SnapshotControls
+          eventManager={props.eventManager}
+          snapshotProps={snapshotProps}
+          state={props.state}
+          onToggleHighlights={_toggleSnapshotHighlights}
+          onStateChange={_changeSnapshotState}
         />
-        <ScriptError error={scriptError} />
-        <div className='cover' />
-      </div>
-    )
-  }
-
-  componentDidMount () {
-    const config = this.props.config
-
-    this.autIframe = new AutIframe(config)
+      ),
+    }),
+  )
 
-    this.props.eventManager.on('visit:failed', this.autIframe.showVisitFailure)
-    this.props.eventManager.on('before:screenshot', this.autIframe.beforeScreenshot)
-    this.props.eventManager.on('after:screenshot', this.autIframe.afterScreenshot)
-    this.props.eventManager.on('script:error', this._setScriptError)
+  useEffect(() => {
+    props.eventManager.on('visit:failed', autIframe.showVisitFailure)
+    props.eventManager.on('before:screenshot', autIframe.beforeScreenshot)
+    props.eventManager.on('after:screenshot', autIframe.afterScreenshot)
+    props.eventManager.on('script:error', _setScriptError)
 
     // TODO: need to take headless mode into account
     // may need to not display reporter if more than 200 tests
-    this.props.eventManager.on('restart', () => {
-      this._run(this.props.state.spec, config)
+    props.eventManager.on('restart', () => {
+      _run(props.state.spec, props.config)
     })
 
-    this.props.eventManager.on('print:selector:elements:to:console', this._printSelectorElementsToConsole)
+    props.eventManager.on('print:selector:elements:to:console', autIframe.printSelectorElementsToConsole)
+
+    props.eventManager.start(props.config)
+
+    setDisposers([
+      autorun(() => {
+        autIframe.toggleSelectorPlayground(selectorPlaygroundModel.isEnabled)
+      }),
+      autorun(() => {
+        autIframe.toggleSelectorHighlight(selectorPlaygroundModel.isShowingHighlight)
+      }),
+      autorun(() => {
+        if (props.state.spec) {
+          _run(props.state.spec, props.config)
+        }
+      }),
+    ])
```

Comment:
`disposers` isn't actually used outside of this hook. You're just using it to represent all of the things you need to clean up. Just assign this array to a local var and call `dispose` on them in the cleanup method (as you're already doing).

Reply:
Using `setState` in this way is forcing the component to rerender _AND_ doesn't even work correctly:

Initial render -> run effect + setState -> render again

After some time

`componentWillUnmount` -> run effect cleanup -> run `disposers.forEach` on the reference of `disposers` that existed _when the hook first ran_. Since the `useEffect` has not run since the initial render, `disposers` is an empty array, and nothing gets disposed.

----

It's important to remember that the dependency arrays in React hooks are not simply "rerun when one of these changes". They define what values in the closure are allowed to be updated, so if a value doesn't appear in the dependency array, that value never changes in the hook.

Reply:
Right, this makes perfect sense. I am not sure what I was thinking with `useState` here.